src/aws-api-mcp-server/awslabs/aws_api_mcp_server/core/security/policy.py

```diff
@@ -0,0 +1,190 @@
+
+    def get_decision(
+        self, service: str, operation: str, is_read_only: bool, supports_elicitation: bool
+    ) -> PolicyDecision:
+        """Get policy decision for a service/operation combination.
+
+        Priority: deny > elicit > default behavior
+        """
+        operation_kebab = operation.replace('_', '-')
+        operation_kebab = re.sub('([A-Z])', r'-\1', operation_kebab).lower().lstrip('-')
+
+        api_call = f'aws {service} {operation_kebab}'
+
+        # s3 service maps to s3api in CLI for some operations
+        if service == 's3':
+            # For s3 service, also check s3api variant
+            api_call_s3api = f'aws s3api {operation_kebab}'
+
+            # Check denylist first
+            if api_call in self.denylist or api_call_s3api in self.denylist:
+                return PolicyDecision.DENY
+
+            # Check elicit list
+            if api_call in self.elicit_list or api_call_s3api in self.elicit_list:
+                if not supports_elicitation:
+                    return PolicyDecision.DENY
+                return PolicyDecision.ELICIT
+        else:
+            # Check denylist first
+            if api_call in self.denylist:
+                return PolicyDecision.DENY
+
+            # Check elicit list
+            if api_call in self.elicit_list:
+                # If client doesn't support elicitation, treat the elicit list as deny
+                if not supports_elicitation:
+                    return PolicyDecision.DENY
+                return PolicyDecision.ELICIT
+
+        # Default behavior: allow all operations
+        return PolicyDecision.ALLOW
+
+    def check_customization(
+        self, cli_command: str, is_read_only_func, supports_elicitation: bool
+    ) -> Optional[PolicyDecision]:
+        """Check if command matches a customization and return the highest priority decision.
+
+        Returns None if no customization matches.
+        """
+        # Extract base command (e.g., "s3 ls" from "aws s3 ls bucket-name")
+        parts = cli_command.strip().split()
+        if len(parts) < 3 or parts[0] != 'aws':
+            return None
+
+        base_cmd = f'{parts[1]} {parts[2]}'
+
+        if base_cmd not in self.customizations:
+            return None
+
+        api_calls = self.customizations[base_cmd]
+        decisions = []
+
+        # Check the parent command itself
+        parent_api_call = f'aws {base_cmd}'
+        if parent_api_call in self.denylist:
+            return PolicyDecision.DENY
+        elif parent_api_call in self.elicit_list:
+            if not supports_elicitation:
+                return PolicyDecision.DENY
+            decisions.append(PolicyDecision.ELICIT)
+
+        # Check all underlying API calls
+        for api_call in api_calls:
+            # Parse service and operation from api_call
+            api_parts = api_call.strip().split()
+            if len(api_parts) < 3 or api_parts[0] != 'aws':
+                continue
```

Comment:
api_call comes from api_calls (the name is confusing btw, it refers to calls customizations maps to), which is the list we maintain and control. Here we validate if cli command is correctly structured, otherwise we skip it (continue).

I'd say it's better to fail here, or have some kind of customizations file validation in other place. Currently if something is incorrectly configured in our mapping.json we will not know about as it will always be ignored.